### PR TITLE
Search API

### PR DESCRIPTION
**Pull request by Tunous**
https://docs.joinmastodon.org/methods/search/#v2
https://docs.joinmastodon.org/entities/Search/#attributes
https://api.pleroma.social/#operation/SearchController.search2

TODO:
- [x] Example
- [x] Tests
- [x] Documentation
- [x] SwiftLint

### CODE REVIEW

**Comment by Tunous** [comment]
Pleroma has slightly different query parameters. I’m not sure what is the proper way to support them in TootSDK.

File: Tests/TootSDKTests/SearchTests.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
+//  SearchTests.swift
+//  Created by Łukasz Rutkowski on 12/02/2023.
+
+import XCTest
+@testable import TootSDK
+
+final class SearchTests: XCTestCase {
+
+    func testSearchParamsToQueryItems() throws {
```
Comment (kkostov) [inline Tests/TootSDKTests/SearchTests.swift:9]: ❤️

**Comment by kkostov** [comment]
> Pleroma has slightly different query parameters. I’m not sure what is the proper way to support them in TootSDK.

The `search` method can check the current flavour. If `SearchParams` has a value for `exclude_unreviewed` and the current flavour is not `.mastodon`, an error is thrown (you can see some examples of this in `adminGetOauthApps` or `func lookupAccount(uri:)`.

```swift
guard flavour == .mastodon else {
   throw TootSDKError.unsupportedFlavour(current: flavour, required: [. mastodon])
}
```

Regarding the response, `hashtags` will not be of type `Tag` but rather a `TagResult` with only `name` and `url`. 

`with_relationships` can be ignored (as they set a pleroma-specific part of the response which we do not support atm):

**Comment by kkostov** [comment]
Thank you so much for this contribution @Tunous! I also love your attention to detail with including tests and updating the example 🤩. Nicely done!

File: Sources/TootSDK/TootClient/TootClient+Search.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
+//  TootClient+Search.swift
+//  Created by Łukasz Rutkowski on 12/02/2023.
+
+import Foundation
+
+public extension TootClient {
+
+    /// Search for content in accounts, posts and hashtags.
+    /// - Parameter params: The search parameters.
+    func search(_ params: SearchParams) async throws -> Search {
```
[the remark this answers is not in the document]
Comment (Tunous) [reply Sources/TootSDK/TootClient/TootClient+Search.swift]: > if you give me push rights to your branch I can make the tweaks as well

I checked that option so it should work already. I pushed the changes anyway.

**Comment by Tunous** [comment]
> Regarding the response, hashtags will not be of type Tag but rather a TagResult with only name and url.

I believe Tag is correct. Mastodon API shows that tags can return history, which is marked as optional in TootSDK so I think this type should work for Pleroma too.

**Review by kkostov — APPROVED** [review]
🎉!

**Pull request merged**